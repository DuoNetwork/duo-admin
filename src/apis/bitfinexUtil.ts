import ws from 'ws';
import sqlUtil from '../sqlUtil';
import * as CST from '../constants';
import { Trade } from '../types';

export class BitfinexUtil {
	parseTrade(trade: object): Trade {
		const amount: number = parseFloat(trade[2]);
		let trade_type: string = 'buy';
		if (amount > 0) {
			trade_type = 'buy';
		} else {
			trade_type = 'sell';
		}
		return {
			source: CST.EXCHANGE_BITFINEX,
			tradeId: trade[0] + '',
			price: trade[3] + '',
			amount: Math.abs(amount) + '',
			tradeType: trade_type,
			sourceTimestamp: trade[1] + ''
		};
	}

	parseApiResponse(msg: string) {
		let parsedJson = JSON.parse(msg.toString());
		// console.log(parsedJson);
		if (parsedJson != undefined) {
			// handle the snapshot
			if (
				parsedJson.event === undefined &&
				parsedJson[1] != 'hb' &&
				!(parsedJson[1] == 'te' || parsedJson[1] == 'tu')
			) {
				// console.log(parsedJson);
				const snapshotArr = parsedJson[1];
				snapshotArr.forEach(trade => {
					const parsedTrade: Trade = this.parseTrade(trade);
					// console.log(parsedTrade);
					sqlUtil.insertSourceData(parsedTrade);
				});
			} else if (parsedJson[1] != 'hb' && parsedJson[1] == 'te') {
				parsedJson = parsedJson[2];

				const parsedTrade: Trade = this.parseTrade(parsedJson);
				sqlUtil.insertSourceData(parsedTrade);
			}
		}
	}

	// Version 2 WebSocket API ---
	fetchTrades() {
		const w = new ws('wss://api.bitfinex.com/ws/2');

		w.on('message', msg => this.parseApiResponse(msg.toString()));

		const msg = JSON.stringify({
			event: 'subscribe',
			channel: 'trades',
			symbol: 'ETHUSD'
		});

		w.on('open', () => {
			console.log('[Bitfinex]-WebSocket is open');
			w.send(msg);
			console.log('subscribe trade');
		});

		w.on('close', () => {
			console.log('[Bitfinex]-WebSocket is close now');
			console.log('close DB');
		});
	}
}
const bitfinexUtil = new BitfinexUtil();
export default bitfinexUtil;
