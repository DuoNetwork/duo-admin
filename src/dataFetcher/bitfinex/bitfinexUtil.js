'use strict';
const MysqlUtil = require('../../utils/mysqlUtil');
import * as CST from '../../constant';

let dbConn;

const EXCHANGE_NAME = CST.EXCHANGE_BITFINEX;
const DB_HOST = CST.DB_HOST;
const DB_USER = CST.DB_USER;
const DB_PASSWORD = CST.DB_PASSWORD;
const DB_PRICEFEED = CST.DB_PRICEFEED;
const DB_TABLE_TRADE = CST.DB_TABLE_TRADE;

class BitfinexTradeFeedUtil {
	constructor() {
		console.log('begin');
	}

	initDB() {
		console.log('Init the DB');

		mysqlUtil.setup(EXCHANGE_NAME, DB_HOST, DB_USER, DB_PASSWORD, DB_PRICEFEED, DB_TABLE_TRADE);

		mysqlUtil.initDB();
	}

	//Version 2 WebSocket API ---
	fetchETHTradesByOwnWebSocket() {
		const ws = require('ws');
		const w = new ws('wss://api.bitfinex.com/ws/2');

		w.on('message', msg => {
			dbConn = mysqlUtil.dbConn;
			if (dbConn === undefined) {
				bitfinexTradeFeedUtil.initDB();
			}

			var parsedJson = JSON.parse(msg);
			if (parsedJson != undefined) {
				//handle the snopshot
				if (parsedJson.event === undefined && parsedJson[1] != 'hb' && !(parsedJson[1] == 'te' || parsedJson[1] == 'tu')) {
					// console.log(parsedJson);
					var snopshotArr = parsedJson[1];
					snopshotArr.forEach(element => {
						// console.log("===>"+element);
						var amount = parseFloat(element[2]);
						var trade_type = 'buy';
						if (amount > 0) {
							trade_type = 'buy';
						} else {
							trade_type = 'sell';
						}
						// console.log("=>"+trade_type);
						mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, element[0], element[3], Math.abs(amount), trade_type, element[1]);
					});
				} else if (parsedJson[1] != 'hb' && parsedJson[1] == 'te') {
					// console.log("<==="+parsedJson);
					parsedJson = parsedJson[2];
					var amount = parseFloat(parsedJson[2]);
					var trade_type = 'buy';
					if (amount > 0) {
						trade_type = 'buy';
					} else {
						trade_type = 'sell';
					}
					// console.log("=>"+trade_type);
					mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, parsedJson[0], parsedJson[3], Math.abs(amount), trade_type, parsedJson[1]);
				}
			}
		});

		let msg = JSON.stringify({
			event: 'subscribe',
			channel: 'trades',
			symbol: 'ETHUSD'
		});

		w.on('open', () => {
			console.log('[Bitfinex]-WebSocket is open');
			w.send(msg);
			console.log('subscribe trade');
		});

		w.on('close', () => {
			console.log('[Bitfinex]-WebSocket is close now');
			console.log('close DB');
		});
	}
}
let mysqlUtil = new MysqlUtil();
let bitfinexTradeFeedUtil = new BitfinexTradeFeedUtil();
export default bitfinexTradeFeedUtil;
