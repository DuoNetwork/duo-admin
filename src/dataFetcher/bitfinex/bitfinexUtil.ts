import MysqlUtil from '../../utils/MysqlUtil';
import * as CST from '../../constant';

const EXCHANGE_NAME = CST.EXCHANGE_BITFINEX;
const DB_HOST = CST.DB_HOST;
const DB_USER = CST.DB_USER;
const DB_PASSWORD = CST.DB_PASSWORD;
const DB_PRICEFEED = CST.DB_PRICEFEED;
const DB_TABLE_TRADE = CST.DB_TABLE_TRADE;

export class BitfinexTradeFeedUtil {
	mysqlUtil: MysqlUtil;

	constructor() {
		this.mysqlUtil = new MysqlUtil(EXCHANGE_NAME, DB_HOST, DB_USER, DB_PASSWORD, DB_PRICEFEED, DB_TABLE_TRADE);
	}

	initDB() {
		console.log('Init the DB');
		this.mysqlUtil.initDB();
	}

	// Version 2 WebSocket API ---
	fetchETHTradesByOwnWebSocket() {
		const ws = require('ws');
		const w = new ws('wss://api.bitfinex.com/ws/2');

		w.on('message', msg => {
			const dbConn = this.mysqlUtil.dbConn;
			if (dbConn === undefined) {
				this.initDB();
			}

			let parsedJson = JSON.parse(msg);
			if (parsedJson != undefined) {
				// handle the snopshot
				if (parsedJson.event === undefined && parsedJson[1] != 'hb' && !(parsedJson[1] == 'te' || parsedJson[1] == 'tu')) {
					// console.log(parsedJson);
					const snopshotArr = parsedJson[1];
					snopshotArr.forEach(element => {
						// console.log("===>"+element);
						const amount: number = parseFloat(element[2]);
						let trade_type: string = 'buy';
						if (amount > 0) {
							trade_type = 'buy';
						} else {
							trade_type = 'sell';
						}
						// console.log("=>"+trade_type);
						this.mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, element[0], element[3] + '', Math.abs(amount) + '', trade_type, element[1]);
					});
				} else if (parsedJson[1] != 'hb' && parsedJson[1] == 'te') {
					// console.log("<==="+parsedJson);
					parsedJson = parsedJson[2];
					const amount: number = parseFloat(parsedJson[2]);
					let trade_type: string = 'buy';
					if (amount > 0) {
						trade_type = 'buy';
					} else {
						trade_type = 'sell';
					}
					// console.log("=>"+trade_type);
					this.mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, parsedJson[0], parsedJson[3] + '', Math.abs(amount) + '', trade_type, parsedJson[1]);
				}
			}
		});

		const msg = JSON.stringify({
			event: 'subscribe',
			channel: 'trades',
			symbol: 'ETHUSD'
		});

		w.on('open', () => {
			console.log('[Bitfinex]-WebSocket is open');
			w.send(msg);
			console.log('subscribe trade');
		});

		w.on('close', () => {
			console.log('[Bitfinex]-WebSocket is close now');
			console.log('close DB');
		});
	}
}
const bitfinexTradeFeedUtil = new BitfinexTradeFeedUtil();
export default bitfinexTradeFeedUtil;
