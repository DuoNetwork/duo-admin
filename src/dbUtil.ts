import dynamoUtil from './database/dynamoUtil';
import sqlUtil from './database/sqlUtil';
import keyUtil from './keyUtil';
import { IOption, IPrice, ITrade } from './types';
import util from './util';
const localSQLauth = require('./keys/mysql.json');

class DbUtil {
	private dynamo: boolean = false;

	public async init(tool: string, option: IOption) {
		this.dynamo = option.dynamo;
		const role = util.getDynamoRole(tool, option.dynamo);
		const process = util.getStatusProcess(tool, option);
		util.log('role: ' + role);
		util.log('process: ' + process);

		dynamoUtil.init(option.live, role, process);
		if ((['bitfinex', 'gemini', 'kraken', 'gdax', 'commit'].includes(tool) && !option.dynamo) || option.server) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		} else
			sqlUtil.init(localSQLauth.host, localSQLauth.user, localSQLauth.password);
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPrice) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(): Promise<IPrice> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice();
	}

	public readSourceData(currentTimestamp: number): Promise<ITrade[]> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readSourceData(currentTimestamp);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
