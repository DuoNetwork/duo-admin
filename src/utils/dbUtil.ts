import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from '../common/constants';
import { IOption, IPriceFix, ITrade } from '../common/types';
import dynamoUtil from './dynamoUtil';
import keyUtil from './keyUtil';
import sqlUtil from './sqlUtil';
import util from './util';

class DbUtil {
	public dynamo: boolean = false;

	public async init(tool: string, option: IOption, web3Wrapper: Web3Wrapper) {
		this.dynamo = option.dynamo;
		const process = util.getStatusProcess(tool, option);
		util.logInfo('process: ' + process);

		let config = {};
		try {
			config = require('../keys/aws/' + (option.live ? 'live' : 'dev') + '/admin.json');
		} catch (error) {
			util.logError(error);
		}
		await dynamoUtil.init(
			config,
			option.live,
			process,
			(value: string | number) => web3Wrapper.fromWei(value),
			async txHash => {
				const txReceipt = await web3Wrapper.getTransactionReceipt(txHash);
				if (!txReceipt) return null;
				return {
					status: txReceipt.status
				};
			}
		);
		if ([CST.TRADES, CST.COMMIT, CST.CLEAN_DB].includes(tool) && !option.dynamo) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			return sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		}
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPriceFix) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(quote: string, base: string): Promise<IPriceFix> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice(quote, base);
	}

	public readSourceData(
		currentTimestamp: number,
		quote: string,
		base: string
	): Promise<ITrade[]> {
		return this.dynamo
			? Promise.reject('invalid')
			: sqlUtil.readSourceData(currentTimestamp, base, quote);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
