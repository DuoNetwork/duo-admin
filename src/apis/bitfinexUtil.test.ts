import dbUtil from '../dbUtil';
import bitfinexUtil from './bitfinexUtil';
const trades: Array<{ [key: string]: string }> = require('../samples/bitfinex.json');
const apiResponse = JSON.stringify(require('../samples/bitfinexMsg.json'));

test('parseTrade', () => {
	trades.forEach(async trade => {
		const parsedTrade = bitfinexUtil.parseTrade(trade);
		expect(parsedTrade).toMatchSnapshot();
	});
});

test('parseApiResponse', async () => {
	dbUtil.insertSourceData = jest.fn(() => Promise.resolve());
	await bitfinexUtil.parseApiResponse(apiResponse);
	// for (let i = 0; i < 6; i++)
	expect(
		(dbUtil.insertSourceData as jest.Mock<Promise<void>>).mock.calls[0][0]
	).toMatchSnapshot();
});
