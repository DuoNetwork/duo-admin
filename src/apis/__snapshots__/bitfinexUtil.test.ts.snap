// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseApiResponse 1`] = `
Object {
  "amount": 0.70208471,
  "id": "234509508",
  "price": 653.38,
  "source": "BITFINEX",
  "timestamp": 1524809497543,
}
`;

exports[`parseTrade 1`] = `
Object {
  "amount": 0.00671182,
  "id": "234055510",
  "price": 620.31,
  "source": "BITFINEX",
  "timestamp": 1524736613727,
}
`;

exports[`parseTrade 2`] = `
Object {
  "amount": 9.99539725,
  "id": "234055509",
  "price": 620.31,
  "source": "BITFINEX",
  "timestamp": 1524736613725,
}
`;

exports[`parseTrade 3`] = `
Object {
  "amount": 10.00460275,
  "id": "234055496",
  "price": 620.31,
  "source": "BITFINEX",
  "timestamp": 1524736608714,
}
`;

exports[`parseTrade 4`] = `
Object {
  "amount": 0.0069549,
  "id": "234055479",
  "price": 620.31,
  "source": "BITFINEX",
  "timestamp": 1524736604753,
}
`;

exports[`parseTrade 5`] = `
Object {
  "amount": 1.17616107,
  "id": "234055478",
  "price": 620.31,
  "source": "BITFINEX",
  "timestamp": 1524736604751,
}
`;

exports[`parseTrade 6`] = `
Object {
  "amount": 2,
  "id": "234055474",
  "price": 620.3,
  "source": "BITFINEX",
  "timestamp": 1524736603150,
}
`;

exports[`parseTrade 7`] = `
Object {
  "amount": 7,
  "id": "234055473",
  "price": 620.3,
  "source": "BITFINEX",
  "timestamp": 1524736602407,
}
`;
